### ISSUE TITLE
List mailboxes mismatches children flag

### ISSUE THREAD

**Issue by karashenski** [cyrusimap/cyrus-imapd#1843]
OS: OpenSuse Leap 42.1
Cyrus version: 2.5.10, also tested with 2.3.16
Hierarchy separator: /
Use of virtual domains: yes

Steps to reproduce:
# cyradm -u admin@yourdomain.com hostname

> cm office
> cm office/Spam
> lm office*
office (\HasChildren)         office/Spam (\HasNoChildren)

> cm office-sent
> lm office*
office (\HasNoChildren)       office/Spam (\HasNoChildren)  
office-sent (\HasNoChildren)

> lm office
office (\HasChildren)

Children flag is different depending on the existence of another folder which name starts with the same string followed by dash and then something else. It also depends on the lm query. I suggest it is due to lexical sorting since in ASCII minus (-) is before slash (/) - when dash in folder name is replaced by underscore everything is fine.

**Comment by elliefm** [comment]
I've put together a cassandane test to try and reproduce this, but so far haven't been able to: https://github.com/cyrusimap/cassandane/compare/master...elliefm:v25/1843-list-dash-children

I've tried it with both 2.5.10 and the current 2.5 git head, both with and without altnamespace/virtdomains.  It's consistently giving me the expected \HasChildren flag for "office":

```
[snip]
C: 3 create office
S: 3 OK Completed
C: 4 create office/Spam
S: 4 OK Completed
C: 5 subscribe office
S: 5 OK Completed
C: 6 subscribe office/Spam
S: 6 OK Completed
C: 7 list "" "office*"
S: * LIST (\HasChildren) "/" office
S: * LIST (\HasNoChildren) "/" office/Spam
S: 7 OK Completed (0.000 secs 3 calls)
=====> Cyrus::List[117] expect mailbox: office
=====> Cyrus::List[117] expect mailbox: office/Spam
C: 8 create office-sent
S: 8 OK Completed
C: 9 subscribe office-sent
S: 9 OK Completed
C: 10 list "" "office*"
S: * LIST (\HasChildren) "/" office
S: * LIST (\HasNoChildren) "/" office/Spam
S: * LIST (\HasNoChildren) "/" office-sent
S: 10 OK Completed (0.000 secs 4 calls)
=====> Cyrus::List[117] expect mailbox: office
=====> Cyrus::List[117] expect mailbox: office-sent
=====> Cyrus::List[117] expect mailbox: office/Spam
[snip]
```

Is there anything more you can tell us about your environment that might help narrow this down?

**Comment by karashenski** [comment]
The problem was due to the default value (disabled) of **improved_mboxlist_sort** parameter. After enabling it and following the procedure described here https://www.cyrusimap.org/docs/cyrus-imapd/2.5.2/internal/improved_mboxlist_sort.php the problem was solved. Since it is not very uncommon to have mailboxes that contain dash/minus in their name, I would suggest the default value of **improved_mboxlist_sort** to be enabled.

**Comment by elliefm** [comment]
That helps, I think.  Will add some more cassandane tests (with/without improved_mboxlist_sort) to try and reproduce this, but my current hunch is that this may be a cyradm bug, rather than an imapd bug.

**Comment by HGuillemet** [comment]
I had a similar problem with 3.0.11.
A user has the following folders :
"H3"
"H3 something"
"H3/subfolder"
The H3 folder was reported with \HasNoChildren in cyradm, and the subfolder was invisible in the email client he is using (Group Office).
When renaming "H3 something" as "something", "H3/subfolder" reappear in his client.
`improved_mboxlist_sort` is not specified in the configuration files, so I guess it's off.

**Comment by elliefm** [comment]
I wonder if cyradm is suffering from a similar bug to #2877 (i.e. unconditionally performing optimisations that are only valid with improved_mboxlist_sort enabled).

The fix for #2877 was e.g. eadda03e6 (3.0.12)

**Comment by rjbs** [comment]
:robot: I'm closing this issue by automation.

This issue refers only to a long-out-of-support version of Cyrus.  We're trying to get a good handle on our open issues and pull requests, so that we can reliably use them as a backlog of things we hope to work on.  If you are confident that this issue should be reconsidered, reopen it and explain why.

Remember, though, that there are only so many developers working on Cyrus, and we'd like to focus on the most important and/or low-hanging work.

Before you reopen this issue, please check that this issue is still present in the latest release of Cyrus, or the main development branch. An issue is considerably more likely to get resolved if you can provide a pull request which reproduces the issue in a Cassandane test.
